django/test/runner.py

```diff
@@ -601,6 +601,8 @@ def setup_test_environment(self, **kwargs):
     def load_tests_for_label(self, label, discover_kwargs):
         label_as_path = os.path.abspath(label)
         tests = None
+        original_test_name_patterns = self.test_loader.testNamePatterns
+        self.test_loader.testNamePatterns = self.test_name_patterns
```

Comment:
I just realized you need to use a `try-finally` below in case an exception occurs. You might as well put the `self.test_loader._top_level_dir = None` in the `finally` also.

To add a second test for testing the exception case, you can look at `test_load_tests_for_label_file_path` as a model.

Reply:
I noticed that we have early returns in the `load_tests_for_label`, so it will not work in current form. I added tests for all cases and will fix it tomorrow.

Reply:
With three early returns and an exception that can be raised I think we should move it outside of `load_tests_for_label()` :disappointed:

Reply:
Would another option be to define a simple `_load_with_patterns()` context manager and use that in `load_tests_for_label()` around the affected lines? It looks like only `loadTestsFromName()` and `discover()` need to be protected, as some returns don't actually wind up loading.

Reply:
Good idea :+1: